Firestore/Source/Remote/FSTSerializerBeta.mm

```diff
@@ -354,70 +356,74 @@ - (FSTFieldValue *)decodedReferenceValue:(NSString *)resourceName {
               database_id.database_id());
-  return FieldValue::FromReference(_databaseID, key).Wrap();
+  return FieldValue::FromReference(_databaseID, key);
 }
 
-- (GCFSArrayValue *)encodedArrayValue:(FSTArrayValue *)arrayValue {
+- (GCFSArrayValue *)encodedArrayValue:(const FieldValue::Array &)arrayValue {
   GCFSArrayValue *proto = [GCFSArrayValue message];
   NSMutableArray<GCFSValue *> *protoContents = [proto valuesArray];
 
-  [[arrayValue internalValue]
-      enumerateObjectsUsingBlock:^(FSTFieldValue *value, NSUInteger idx, BOOL *stop) {
-        GCFSValue *converted = [self encodedFieldValue:value];
-        [protoContents addObject:converted];
-      }];
+  for (const FieldValue &value : arrayValue) {
+    GCFSValue *converted = [self encodedFieldValue:value];
+    [protoContents addObject:converted];
+  }
   return proto;
 }
 
-- (FSTArrayValue *)decodedArrayValue:(GCFSArrayValue *)arrayValue {
-  NSMutableArray<FSTFieldValue *> *contents =
-      [NSMutableArray arrayWithCapacity:arrayValue.valuesArray_Count];
+- (FieldValue)decodedArrayValue:(GCFSArrayValue *)arrayValue {
+  FieldValue::Array contents;
+  contents.reserve(arrayValue.valuesArray_Count);
+
+  for (GCFSValue *value in arrayValue.valuesArray) {
+    contents.push_back([self decodedFieldValue:value]);
+  }
 
-  [arrayValue.valuesArray
-      enumerateObjectsUsingBlock:^(GCFSValue *value, NSUInteger idx, BOOL *stop) {
-        [contents addObject:[self decodedFieldValue:value]];
-      }];
-  return [[FSTArrayValue alloc] initWithValueNoCopy:contents];
+  return FieldValue::FromArray(std::move(contents));
 }
 
-- (GCFSMapValue *)encodedMapValue:(FSTObjectValue *)value {
+- (GCFSMapValue *)encodedMapValue:(const FieldValue::Map &)value {
   GCFSMapValue *result = [GCFSMapValue message];
-  result.fields = [self encodedFields:value];
+  result.fields = [self encodedMapFields:value];
   return result;
 }
 
-- (FSTObjectValue *)decodedMapValue:(GCFSMapValue *)map {
+- (ObjectValue)decodedMapValue:(GCFSMapValue *)map {
   return [self decodedFields:map.fields];
 }
 
 /**
  * Encodes an FSTObjectValue into a dictionary.
  * @return a new dictionary that can be assigned to a field in another proto.
  */
-- (NSMutableDictionary<NSString *, GCFSValue *> *)encodedFields:(FSTObjectValue *)value {
-  FSTImmutableSortedDictionary<NSString *, FSTFieldValue *> *fields = value.internalValue;
+- (NSMutableDictionary<NSString *, GCFSValue *> *)encodedFields:(const ObjectValue &)value {
+  return [self encodedMapFields:value.GetInternalValue()];
+}
+
+- (NSMutableDictionary<NSString *, GCFSValue *> *)encodedMapFields:(const FieldValue::Map &)value {
   NSMutableDictionary<NSString *, GCFSValue *> *result = [NSMutableDictionary dictionary];
-  [fields enumerateKeysAndObjectsUsingBlock:^(NSString *key, FSTFieldValue *obj, BOOL *stop) {
-    GCFSValue *converted = [self encodedFieldValue:obj];
+
+  for (auto &&kv : value) {
```

Comment:
FWIW, my takeaway from the thread was something like `auto&& might avoid some corner cases in the case where a) the code in question is in a template function, and the type of loop variable depends on the template parameter; b) the loop variable is being modified`. Basically, `auto&` would fail to bind to a temporary, while `auto&&` would not. Another argument was that `auto&&` isn't really worse, with which I partially disagree because I prefer having explicit `const` where appropriate.

I don't think the argument above applies here at all (it's not a template function). I guess one could argue that `auto&&` style could be used everywhere for consistency even when its exact behavior doesn't matter, but I'm not convinced -- I feel like explaining what `auto&&` does (basically, what forwarding references do) is harder than explaining what `const auto&` does, so this seems to me like a style that is not particularly friendly to less C++-oriented people, while the benefits seem somewhat esoteric.

Reply:
`const auto &` works for me.